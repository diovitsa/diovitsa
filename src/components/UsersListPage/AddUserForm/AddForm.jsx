import React from 'react';
import { AddFormStyles } from './styles';
import withStyles from '@material-ui/core/styles/withStyles';
import TextField from '@material-ui/core/TextField';
import { ButtonStyles } from './styles';
import Button from '@material-ui/core/Button/index';

const CustomizedButton = withStyles(ButtonStyles)(Button);


class AddForm extends React.Component {

  state = {
    userName: '',
    userEmail: '',
    password: '',
  };

  componentDidMount() {
    document.addEventListener('keydown', this.keydown, false);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.keydown, false);
  }

  keydown = (e) => {
    if (e.key === 'Enter' && (e.target === this.nameInput || e.target === this.emailInput || e.target === this.passwordInput)) {
      this.addUser();
    }
  };

  nameInputHandler(e) {
    this.nameInput = e.target;
    this.setState({ userName: e.target.value });
  };

  emailInputHandler(e) {
    this.emailInput = e.target;
    this.setState({ userEmail: e.target.value });
  };

  passwordInputHandler(e) {
    this.passwordInput = e.target;
    this.setState({ password: e.target.value });
  }

  clearInputValues() {
    if (this.nameInput && this.emailInput && this.passwordInput) {
      this.nameInput.value = '';
      this.emailInput.value = '';
      this.passwordInput.value = '';
    }
  }

  addUser() {
    const { userName, userEmail, password } = this.state;

    this.clearInputValues();
    this.props.onUserAdd({ 'password': password, 'email': userEmail, 'name': userName });
  }

  render() {
    const { root, input } = this.props.classes;
    return (
      <div className={root}>
        <TextField className={input} label='Name' type="text" onChange={(e) => this.nameInputHandler(e)}/>
        <TextField className={input} label="Email" type="text" onChange={(e) => this.emailInputHandler(e)}/>
        <TextField className={input} label="Password" type="password" onChange={(e) => this.passwordInputHandler(e)}/>
        <CustomizedButton onClick={() => this.addUser()} variant="contained" color="primary">Add user</CustomizedButton>
      </div>
    );
  }
}

export default withStyles(AddFormStyles)(AddForm);