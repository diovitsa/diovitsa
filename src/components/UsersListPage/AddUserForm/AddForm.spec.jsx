import React from 'react';
import { shallow } from 'enzyme';
import AddForm from './AddForm';

describe('LogInForm', () => {
  let component;
  let onUserAdd;

  function getComponent() {
    const testingComponent = shallow(<AddForm onUserAdd={onUserAdd}/>).dive();
    return testingComponent.instance();
  }

  beforeEach(() => {
    onUserAdd = jest.fn();
    component = getComponent();
  });

  it('should render properly', () => {
    expect(getComponent()).toMatchSnapshot();
  });

  describe('componentDidMount', () => {

    beforeEach(() => {
      jest.spyOn(document, 'addEventListener');
      component.componentDidMount();
    });

    it('should init keydown listener', () => {
      expect(document.addEventListener).toHaveBeenCalledWith('keydown', component.keydown, false);
    });

  });

  describe('componentWillUnmount', () => {

    beforeEach(() => {
      jest.spyOn(document, 'removeEventListener');
      component.componentWillUnmount();
    });

    it('should remove keydown listener', () => {
      expect(document.removeEventListener).toHaveBeenCalledWith('keydown', component.keydown, false);
    });

  });

  describe('keydown', () => {

    beforeEach(() => {
      jest.spyOn(component, 'addUser');
    });

    describe('when email or password input is on focus', () => {

      beforeEach(() => {
        const event = {
          key: 'Enter',
          target: 'passwordInput'
        };
        component.passwordInput = 'passwordInput';
        component.keydown(event);
      });

      it('should call signIn method', () => {
        expect(component.addUser).toHaveBeenCalled();
      });
    });

    describe('when key it\'s not Enter', () => {

      beforeEach(() => {
        const event = {
          key: 'Space',
          target: 'passwordInput'
        };

        component.passwordInput = 'passwordInput';
        component.keydown(event);
      });

      it('should call addUser method', () => {
        expect(component.addUser).not.toHaveBeenCalled();
      });
    });
  });

  describe('addUser', () => {

    beforeEach(() => {
      jest.spyOn(component, 'clearInputValues');
      component.state = { userEmail: 'testEmail', password: 'testPassword', userName: 'testName' };
      component.addUser();
    });

    it('should clear all input values', () => {
      expect(component.clearInputValues).toHaveBeenCalled();
    });

    it('should call component.props.onSignIn', () => {
      expect(component.props.onUserAdd).toHaveBeenCalledWith({
        'email': 'testEmail',
        'name': 'testName',
        'password': 'testPassword'
      });
    });
  });

  describe('clearInputValues', () => {

    beforeEach(() => {
      component.emailInput = { value: 'email@gmail.com' };
      component.nameInput = { value: 'testName' };
      component.passwordInput = { value: 'testPassword' };

      component.clearInputValues();
    });

    it('should clear all input values', () => {
      expect(component.emailInput.value).toEqual('');
      expect(component.nameInput.value).toEqual('');
      expect(component.passwordInput.value).toEqual('');
    });

  });

});
