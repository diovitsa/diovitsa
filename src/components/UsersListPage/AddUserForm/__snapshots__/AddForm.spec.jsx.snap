// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogInForm should render properly 1`] = `
AddForm {
  "context": Object {},
  "keydown": [Function],
  "props": Object {
    "classes": Object {
      "input": "AddForm-input-2",
      "root": "AddForm-root-1",
    },
    "onUserAdd": [MockFunction],
  },
  "refs": Object {},
  "setState": [Function],
  "state": Object {
    "password": "",
    "userEmail": "",
    "userName": "",
  },
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useState": [Function],
      },
      "_element": <AddForm
        classes={
          Object {
            "input": "AddForm-input-2",
            "root": "AddForm-root-1",
          }
        }
        onUserAdd={[MockFunction]}
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <div
        className="AddForm-root-1"
      >
        <TextField
          className="AddForm-input-2"
          label="Name"
          onChange={[Function]}
          required={false}
          select={false}
          type="text"
          variant="standard"
        />
        <TextField
          className="AddForm-input-2"
          label="Email"
          onChange={[Function]}
          required={false}
          select={false}
          type="text"
          variant="standard"
        />
        <TextField
          className="AddForm-input-2"
          label="Password"
          onChange={[Function]}
          required={false}
          select={false}
          type="password"
          variant="standard"
        />
        <WithStyles(WithStyles(Button))
          color="primary"
          onClick={[Function]}
          variant="contained"
        >
          Add user
        </WithStyles(WithStyles(Button))>
      </div>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
  Symbol(enzyme.__setState__): [Function],
}
`;
